File: android/app/src/main/java/com/neomorphapp/NeumorphViewManager.java
package com.neomorphapp;

import android.graphics.Color;
import android.util.Log;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;

import javax.annotation.Nullable;

import soup.neumorphism.CornerFamily;
import soup.neumorphism.NeumorphCardView;
import soup.neumorphism.NeumorphShapeAppearanceModel;
import soup.neumorphism.NeumorphShapeAppearanceModel.Builder;
import soup.neumorphism.ShapeType;

public class NeumorphViewManager extends ViewGroupManager<NeumorphCardView> {

    public static final String REACT_CLASS = "RNCNeumorph";
    public int shapeType = ShapeType.DEFAULT;
    public NeumorphShapeAppearanceModel borderRadius = new Builder()
		.setAllCorners(CornerFamily.OVAL)
		.setAllCornerSizes(20f)
		.build();

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected NeumorphCardView createViewInstance(ThemedReactContext reactContext) {
        NeumorphCardView neumorphCardView = new NeumorphCardView(reactContext);
//        neumorphCardView.setShadowElevation(12);
//        neumorphCardView.setShadowColorLight(0x60FFFFFF);
//        neumorphCardView.setShadowColorDark(0x60000000);
//        neumorphCardView.setBackgroundColor(0x212121);
        neumorphCardView.setStrokeWidth(10);
        neumorphCardView.setShapeAppearanceModel(borderRadius);
        neumorphCardView.invalidate();
        return  neumorphCardView;
    }

    @ReactProp(name = "backgroundColor")
    public void setBackgroundColor(NeumorphCardView neumorphCardView, String backgroundColor) {
    	if(backgroundColor.equals("transparent")) {
    		neumorphCardView.setBackgroundColor(0x60FFFFFF);
		} else {
			neumorphCardView.setBackgroundColor(Color.parseColor(backgroundColor));
		}
    }

    @ReactProp(name = "inner")
    public void setTypeShadow(NeumorphCardView neumorphCardView, Boolean inner) {
        if(inner) {
            neumorphCardView.setShapeType(ShapeType.PRESSED);
        } else {
        	neumorphCardView.setShapeType(ShapeType.FLAT);
		}
    }

    @ReactProp(name = "basin")
	public void setShapeType(NeumorphCardView neumorphCardView, Boolean type) {
    	if(type) {
			neumorphCardView.setShapeType(ShapeType.BASIN);
		} else {
    		neumorphCardView.setShapeType(ShapeType.DEFAULT);
		}
	}

	@ReactProp(name = "darkShadowColor")
	public void setShadowDark(NeumorphCardView neumorphCardView, String darkShadowColor) {
		neumorphCardView.setShadowColorDark(Color.parseColor(darkShadowColor));
	}

	@ReactProp(name = "lightShadowColor")
	public void setShadowLight(NeumorphCardView neumorphCardView, String lightShadowColor) {
		neumorphCardView.setShadowColorLight(Color.parseColor(lightShadowColor));
	}

	@ReactProp(name = "shadowOffset")
	public void setShadowOffset(NeumorphCardView neumorphCardView, ReadableMap shadowOffset) {
    	neumorphCardView.setInset(
    		shadowOffset.getInt("width"),
			shadowOffset.getInt("height"),
			shadowOffset.getInt("width"),
			shadowOffset.getInt("height")
		);
	}

	@ReactProp(name = "borderRadius", defaultInt = 0)
	public void setBorderRadius(NeumorphCardView neumorphCardView, Integer borderRadius) {
		NeumorphShapeAppearanceModel newBorderRadius = new Builder()
			.setAllCornerSizes(borderRadius)
			.build();
		neumorphCardView.setShapeAppearanceModel(newBorderRadius);
	}

	@ReactProp(name = "shadowRadius", defaultInt = 0)
	public void setShadowRadius(NeumorphCardView neumorphCardView, Integer shadowRadius) {
    	neumorphCardView.setShadowElevation(shadowRadius);
	}

}

File: android/app/src/main/java/com/neomorphapp/NeumorphViewManager.kt
package com.neomorphapp

import android.graphics.Color
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.views.view.ReactViewGroup
import soup.neumorphism.CornerFamily
import soup.neumorphism.NeumorphCardView
import soup.neumorphism.NeumorphShapeAppearanceModel
import soup.neumorphism.ShapeType

class NeumorphViewManager : ViewGroupManager<ReactViewGroup>() {
    companion object {
        const val REACT_CLASS = "RNCNeumorph"
    }

    override fun getName(): String {
        return REACT_CLASS
    }

    override fun createViewInstance(reactContext: ThemedReactContext): ReactViewGroup {
        val neumorphCardView = NeumorphCardView(reactContext)
        return neumorphCardView
    }

    @ReactProp(name = "backgroundColor")
    fun setBackgroundColor(neumorphCardView: NeumorphCardView, backgroundColor: String) {
        if (backgroundColor !== "transparent") {
            neumorphCardView.setBackgroundColor(Color.parseColor(backgroundColor))
        }
    }

    @ReactProp(name = "inner")
    fun setTypeShadow(neumorphCardView: NeumorphCardView, inner: Boolean) {
        if(inner) {
            neumorphCardView.setShapeType(ShapeType.PRESSED);
        } else {
        	neumorphCardView.setShapeType(ShapeType.FLAT);
		}
    }

    @ReactProp(name = "basin")
    fun setShapeType(neumorphCardView: NeumorphCardView, type: Boolean) {
    	if(type) {
			neumorphCardView.setShapeType(ShapeType.BASIN);
		} else {
    		neumorphCardView.setShapeType(ShapeType.DEFAULT);
		}
    }

    @ReactProp(name = "darkShadowColor")
    fun setShadowDark(neumorphCardView: NeumorphCardView, darkShadowColor: ReadableMap) {
    	val alpha = darkShadowColor.getInt("alpha")
    	val r = darkShadowColor.getInt("r")
    	val g = darkShadowColor.getInt("g")
    	val b = darkShadowColor.getInt("b")
		neumorphCardView.setShadowColorDark(Color.argb(alpha, r, g, b));
    }

    @ReactProp(name = "lightShadowColor")
    fun setShadowLight(neumorphCardView: NeumorphCardView, lightShadowColor: ReadableMap) {
		val alpha = lightShadowColor.getInt("alpha")
		val r = lightShadowColor.getInt("r")
		val g = lightShadowColor.getInt("g")
		val b = lightShadowColor.getInt("b")
		neumorphCardView.setShadowColorLight(Color.argb(alpha, r, g, b));
    }

    @ReactProp(name = "borderRadius")
    fun setBorderRadiusTwo(neumorphCardView: NeumorphCardView, borderRadius: ReadableMap) {
        val type = borderRadius.getBoolean("type");
        val radius = borderRadius.getInt("radius");
        val newBorderRadius: NeumorphShapeAppearanceModel;
        if(type) {
            neumorphCardView.setShapeAppearanceModel(
                    NeumorphShapeAppearanceModel.builder().setAllCorners(CornerFamily.OVAL).build()
            )
        } else {
            neumorphCardView.setShapeAppearanceModel(
                    NeumorphShapeAppearanceModel.builder().setAllCorners(
                            CornerFamily.ROUNDED, radius.toFloat()
                    ).build()
            )
        }
    }

    @ReactProp(name = "swapShadow", defaultBoolean = false)
    fun setSwapShadow(neumorphCardView: NeumorphCardView, swapShadow: Boolean) {
        neumorphCardView.setSwapShadow(swapShadow)
    }

    @ReactProp(name = "shadowRadius", defaultFloat = 12f)
    fun setShadowRadius(neumorphCardView: NeumorphCardView, shadowRadius: Float) {
        neumorphCardView.setShadowElevation(shadowRadius * 2)
    }
}
